packages/target-tools/src/telemetryProvider.js

```diff
@@ -39,71 +51,82 @@ export function TelemetryProvider(
       return;
     }
 
-    const timestamp = now();
-
-    telemetryEntries.push({
+    telemetryDao.addEntry({
       requestId: renderId,
-      timestamp,
+      timestamp: now(),
       execution
     });
   }
 
+  function addRequestEntry(requestId, entry) {
+    telemetryDao.addEntry({
+      requestId,
+      timestamp: now(),
+      ...entry
+    });
+  }
+
   /**
    * @param {import("@adobe/target-tools/delivery-api-client/models/TelemetryEntry").TelemetryEntry} entry
    */
-  function addEntry(request, entry, status, decisioningMethod = method) {
+  function addArtifactRequestEntry(requestId, entry) {
     if (!telemetryEnabled || !entry) {
       return;
     }
+    addRequestEntry(requestId, entry);
+  }
 
-    const { requestId } = request;
-    const timestamp = now();
+  /**
+   * @param {import("@adobe/target-tools/delivery-api-client/models/TelemetryEntry").TelemetryEntry} entry
+   */
+  function addDeliveryRequestEntry(
+    request,
+    entry,
+    status,
+    decisioningMethod = method
+  ) {
+    if (!telemetryEnabled || !entry) {
+      return;
+    }
 
-    telemetryEntries.push({
-      requestId,
-      timestamp,
-      mode: getMode(status, decisioningMethod),
-      features: {
-        decisioningMethod,
-        executePageLoad: isExecutePageLoad(request),
-        executeMboxCount: executeMboxCount(request),
-        prefetchPageLoad: isPrefetchPageLoad(request),
-        prefetchMboxCount: prefetchMboxCount(request),
-        prefetchViewCount: prefetchViewCount(request)
-      },
-      ...entry
-    });
-  }
+    entry.mode = getMode(status, decisioningMethod);
+    entry.features = {
+      decisioningMethod,
+      executePageLoad: isExecutePageLoad(request),
+      executeMboxCount: executeMboxCount(request),
+      prefetchPageLoad: isPrefetchPageLoad(request),
+      prefetchMboxCount: prefetchMboxCount(request),
+      prefetchViewCount: prefetchViewCount(request)
+    };
 
-  function getEntries() {
-    return telemetryEntries;
+    const { requestId } = request;
+
+    addRequestEntry(requestId, entry);
   }
 
-  function clearEntries() {
-    telemetryEntries = [];
+  function getAndClearEntries() {
+    return telemetryDao.getAndClearEntries();
   }
 
   function hasEntries() {
-    return telemetryEntries.length !== 0;
+    return telemetryDao.hasEntries();
   }
 
   function executeTelemetries(deliveryRequest) {
```

Comment:
You have been focused on telemetry for some time so it probably doesn't phase you, but the `executeTelemetries` name is not obvious what it does without inspecting the method.  My  mind thinks "this function is sending the telemetry entries to delivery api".  But it's actually just transferring telemetry entries to the request in preparation for that.  Can we change it to be more clear?  Maybe a name like `transferTelemetryEntriesToDeliveryRequest`?  Not as short, but more readable.  I noticed you have an `executeTelemetries` method passed into the provider which is named `addTelemetryToDeliveryRequest`.  Why is `executeTelemetries` part of utils.js and passed into the `TelemetryProvider`?  What is the benefit of that over having it all be part of TelemetryProvider itself? It seems like a small refactor here could make the code more readable and all in one place?

Reply:
This is remnants of an older design we had in place at the beginning of the project.  I'll clean it up since it's just confusing now.